### PR TITLE
Phase 2 Goal 4 - WL LOCA Integration

### PR DESCRIPTION
**Pull request by claalmve**
# Description of PR

**Summary of changes and related issue**
This PR includes LOCA2 data as an available option for a warming levels approach for the `cava_data` function.

**Relevant motivation and context**
For phase 2 goal 4 of SCE pilot development.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by vicford — CHANGES_REQUESTED** [review]
Getting a pieces mismatch error on trying to do LOCA2 for the time based likely seasonal event and 1-in-X examples @claalmve

**Review by vicford — CHANGES_REQUESTED** [review]
Good to go:
- Likely seasonal events (time)
- Likely seasonal event (GWL based!!)
- Heat index failing

Errors:
- 1-in-x calculating metric failed time-based (off-ramp was fine)
```
ValueError                                Traceback (most recent call last)
Cell In [6], line 1
----> 1 data = cava_data(
      2     ## Set-up
      3     example_locs.iloc[:1],
      4     time_start_year=2070,
      5     time_end_year=2090,
      6     downscaling_method="Statistical",  # WRF data 
      7     approach="time",  
      8     ssp_data=["SSP3-7.0"],
      9     
     10     ## 1-in-X event specific arguments
     11     variable="Air Temperature at 2m",
     12     metric_calc="max", # daily maximum temperature
     13     one_in_x=10, # One-in-X
     14     units="degF", # change units
     15     
     16     ## Export
     17     export_method="both",
     18     file_format="NetCDF",
     19 )

File ~/src/climakitae/climakitae/explore/vulnerability.py:564, in cava_data(input_locations, variable, approach, downscaling_method, time_start_year, time_end_year, historical_data, ssp_data, warming_level, metric_calc, heat_idx_threshold, one_in_x, percentile, season, units, wrf_bias_adjust, export_method, separate_files, file_format)
    562     # Calculate and export into separate files
    563     for point in loaded_data:
--> 564         calc_val = _metric_agg(
    565             point,
    566             approach,
    567             metric_calc,
    568             heat_idx_threshold,
    569             one_in_x,
    570             percentile,
    571             names["var_name"],
    572         )
    573         metric_data.append(calc_val)
    575 # Export a combined file of all metrics
    576 else:

File ~/src/climakitae/climakitae/explore/vulnerability.py:144, in _metric_agg(da, approach, metric, heat_idx_threshold, one_in_x, percentile, name_of_calc)
    141 if approach == "warming_level":
    142     da = add_dummy_time_to_wl(da)
--> 144 metric_data = _apply_metric(da)
    145 return metric_data

File ~/src/climakitae/climakitae/explore/vulnerability.py:113, in _metric_agg.<locals>._apply_metric(da)
    102 for (
    103     sim
    104 ) in (
    105     da.simulation.values
    106 ):  # Can use xr.apply_ufunc, not just a for loop here
    107     ams = get_block_maxima(
    108         da.sel(simulation=sim),
    109         extremes_type=metric,
    110         groupby=(1, "day"),
    111         check_ess=False,
    112     )
--> 113     return_values = get_return_value(
    114         ams, return_period=one_in_x, multiple_points=False
    115     )["return_value"]
    116     return_vals.append(return_values)
    117 calc_val = xr.concat(return_vals, dim="simulation")

File ~/src/climakitae/climakitae/explore/threshold_tools.py:737, in get_return_value(bms, return_period, distr, bootstrap_runs, conf_int_lower_bound, conf_int_upper_bound, multiple_points)
    703 def get_return_value(
    704     bms,
    705     return_period=10,
   (...)
    710     multiple_points=True,
    711 ):
    712     """Creates xarray Dataset with return values and confidence intervals from maximum series.
    713 
    714     Parameters
   (...)
    734         Dataset with return values and confidence intervals
    735     """
--> 737     return _get_return_variable(
    738         bms,
    739         "return_value",
    740         return_period,
    741         distr,
    742         bootstrap_runs,
    743         conf_int_lower_bound,
    744         conf_int_upper_bound,
    745         multiple_points,
    746     )

File ~/src/climakitae/climakitae/explore/threshold_tools.py:659, in _get_return_variable(bms, data_variable, arg_value, distr, bootstrap_runs, conf_int_lower_bound, conf_int_upper_bound, multiple_points)
    646     conf_int_lower_limit, conf_int_upper_limit = _conf_int(
    647         bms=bms,
    648         distr=distr,
   (...)
    654         block_size=block_size,
    655     )
    657     return return_variable, conf_int_lower_limit, conf_int_upper_limit
--> 659 return_variable, conf_int_lower_limit, conf_int_upper_limit = xr.apply_ufunc(
    660     _return_variable,
    661     bms,
    662     input_core_dims=[["time"]],
    663     exclude_dims=set(("time",)),
    664     output_core_dims=[[], [], []],
    665 )
    667 return_variable = return_variable.rename(data_variable)
    668 new_ds = return_variable.to_dataset()

File /srv/conda/envs/notebook/lib/python3.9/site-packages/xarray/core/computation.py:1204, in apply_ufunc(func, input_core_dims, output_core_dims, exclude_dims, vectorize, join, dataset_join, dataset_fill_value, keep_attrs, kwargs, dask, output_dtypes, output_sizes, meta, dask_gufunc_kwargs, *args)
   1202 # feed DataArray apply_variable_ufunc through apply_dataarray_vfunc
   1203 elif any(isinstance(a, DataArray) for a in args):
-> 1204     return apply_dataarray_vfunc(
   1205         variables_vfunc,
   1206         *args,
   1207         signature=signature,
   1208         join=join,
   1209         exclude_dims=exclude_dims,
   1210         keep_attrs=keep_attrs,
   1211     )
   1212 # feed Variables directly through apply_variable_ufunc
   1213 elif any(isinstance(a, Variable) for a in args):

File /srv/conda/envs/notebook/lib/python3.9/site-packages/xarray/core/computation.py:315, in apply_dataarray_vfunc(func, signature, join, exclude_dims, keep_attrs, *args)
    310 result_coords, result_indexes = build_output_coords_and_indexes(
    311     args, signature, exclude_dims, combine_attrs=keep_attrs
    312 )
    314 data_vars = [getattr(a, "variable", a) for a in args]
--> 315 result_var = func(*data_vars)
    317 out: tuple[DataArray, ...] | DataArray
    318 if signature.num_outputs > 1:

File /srv/conda/envs/notebook/lib/python3.9/site-packages/xarray/core/computation.py:771, in apply_variable_ufunc(func, signature, exclude_dims, dask, output_dtypes, vectorize, keep_attrs, dask_gufunc_kwargs, *args)
    766     if vectorize:
    767         func = _vectorize(
    768             func, signature, output_dtypes=output_dtypes, exclude_dims=exclude_dims
    769         )
--> 771 result_data = func(*input_data)
    773 if signature.num_outputs == 1:
    774     result_data = (result_data,)

File ~/src/climakitae/climakitae/explore/threshold_tools.py:646, in _get_return_variable.<locals>._return_variable(bms)
    643 except (ValueError, ZeroDivisionError):
    644     return_variable = np.nan
--> 646 conf_int_lower_limit, conf_int_upper_limit = _conf_int(
    647     bms=bms,
    648     distr=distr,
    649     data_variable=data_variable,
    650     arg_value=arg_value,
    651     bootstrap_runs=bootstrap_runs,
    652     conf_int_lower_bound=conf_int_lower_bound,
    653     conf_int_upper_bound=conf_int_upper_bound,
    654     block_size=block_size,
    655 )
    657 return return_variable, conf_int_lower_limit, conf_int_upper_limit

File ~/src/climakitae/climakitae/explore/threshold_tools.py:562, in _conf_int(bms, distr, data_variable, arg_value, bootstrap_runs, conf_int_lower_bound, conf_int_upper_bound, block_size)
    559 bootstrap_values = []
    561 for _ in range(bootstrap_runs):
--> 562     result = _bootstrap(bms, distr, data_variable, arg_value, block_size)
    563     bootstrap_values.append(result)
    565 conf_int_array = np.percentile(
    566     bootstrap_values, [conf_int_lower_bound, conf_int_upper_bound]
    567 )

File ~/src/climakitae/climakitae/explore/threshold_tools.py:510, in _bootstrap(bms, distr, data_variable, arg_value, block_size)
    507 distr_func = _get_distr_func(distr)
    509 sample_size = len(bms)
--> 510 new_bms = np.random.choice(bms, size=sample_size, replace=True)
    512 try:
    513     parameters, fitted_distr = _get_fitted_distr(new_bms, distr, distr_func)

File mtrand.pyx:911, in numpy.random.mtrand.RandomState.choice()

ValueError: a must be 1-dimensional
```

Caveat: this was before the reconstruction data PR was merged in, so maybe this is changed

**Comment by claalmve** [comment]
- [ ] likely WL
- [ ] likely time
- [ ] 1 in X wl
- [ ] 1 in X time
- [ ] heat index WL
- [ ] heat index time

**Comment by vicford** [comment]
@claalmve I think the pdb debugger is still in the code because it is popping up for LOCA2 1-in-X time
Screenshot 2024-06-28 at 11 02 04 AM

File: climakitae/explore/vulnerability.py [hunk 1]
```diff
@@ -91,6 +97,9 @@ def _apply_metric(da):
 
         elif one_in_x:  # Calculate 1-in-X events
 
+            import pdb
+
+            pdb.set_trace()
```
Comment (vicford) [inline climakitae/explore/vulnerability.py:102]:
```suggestion

```

**Comment by claalmve** [comment]
> @claalmve I think the pdb debugger is still in the code because it is popping up for LOCA2 1-in-X time Screenshot 2024-06-28 at 11 02 04 AM

🤦 I apologize! I forgot to push the latest code. I removed it and without the debugger, it should run just fine now!

**Review by vicford — APPROVED** [review]
1-in-X now working for LOCA for both time based and WL approaches! I will update the notebook to remove the "in progress piece" and add 2.5 as an option too

**Pull request merged**